beacon/goclient/events.go

```diff
@@ -0,0 +1,88 @@
+package goclient
+
+import (
+	"context"
+
+	apiv1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type event interface {
+	*apiv1.HeadEvent
+}
+
+type subscriber[T event] struct {
+	Identifier string
+	Channel    chan T
+}
+
+func (gc *GoClient) SubscribeToHeadEvents(ctx context.Context, subscriberIdentifier string) (<-chan *apiv1.HeadEvent, error) {
+	gc.subscribersLock.Lock()
+	defer gc.subscribersLock.Unlock()
+	logger := gc.log.With(zap.String("subscriber_identifier", subscriberIdentifier))
+
+	logger.Info("adding 'head' event subscriber")
+
+	if len(gc.headEventSubscribers) == 0 {
+		logger.Info("launching event listener")
+		if err := gc.startEventListener(ctx); err != nil {
+			const errMsg = "failed to start event listener"
+			logger.Error(errMsg, zap.Error(err))
+			return nil, errors.Wrap(err, errMsg)
+		}
```

Comment:
I don't understand why we need to do this "lazy" starting of the event listener. we're always going to subscribe so its better to just start it on the ctor.

Reply:
I do see what you mean. But even without the blockrootcache, there's DutyScheduler listening on these events. It seems like there's no foreseen situation where we can remove all the subscriptions, so I don't see a lot of advantage to preparing for such situation before we can even imagine why would it happen. I'm not saying it doesn't work as it is, it just adds complexity and tangles the code a bit too much IMO for no good reason.

we can use some more input here - @MatusKysel @moshe-blox

Reply:
@oleg-ssvlabs if we decide to drop weighted attestation data, we'd still need to subscribe to head events for DutyScheduler

and if we decide to drop that as well, we can just empty supportedTopics slice😄

should be a simpler change than modifying the code, so perhaps the condition should be if `len(supportedTopics) > 0` then start listening